crates/editor/src/rainbow_brackets.rs

```diff
@@ -0,0 +1,286 @@
+use crate::Editor;
+use gpui::{Context, Hsla, Window};
+use language::Bias;
+use std::collections::HashMap;
+use std::ops::Range;
+use text::{Anchor, ToOffset};
+
+/// Rainbow bracket highlighting uses multiple colors to distinguish bracket nesting levels
+pub fn refresh_rainbow_bracket_highlights(
+    editor: &mut Editor,
+    _window: &mut Window,
+    cx: &mut Context<Editor>,
+) {
+    // Clear existing rainbow highlights for all levels
+    clear_current_rainbow_highlights(editor, cx);
+
+    let multi_buffer = editor.buffer().read(cx);
+    let multi_buffer_snapshot = multi_buffer.snapshot(cx);
+
+    // For now, handle only singleton buffers
+    if let Some((_, _, buffer_snapshot)) = multi_buffer_snapshot.as_singleton() {
+        let language = buffer_snapshot.language();
+
+        if let Some(language) = language {
+            if let Some(grammar) = language.grammar() {
+                if let Some(rainbow_config) = &grammar.rainbow_config {
```

Comment:
Could be simplified using `and_then` to avoid 3 level of nesting.

```suggestion
        if let Some(rainbow_config) = language
            .and_then(|lang| lang.grammar())
            .and_then(|grammar| grammar.rainbow_config.as_ref())
```